MFRC522.cpp

```diff
@@ -1870,6 +2077,34 @@ bool MFRC522::PICC_IsNewCardPresent() {
  * @return bool
  */
 bool MFRC522::PICC_ReadCardSerial() {
-	MFRC522::StatusCode result = PICC_Select(&uid);
-	return (result == STATUS_OK);
+	MFRC522::StatusCode result = PICC_Select(&card.uid);
+
+	// Backward compatibility
+	uid.size = card.uid.size;
+	uid.sak = card.uid.sak;
+	memcpy(uid.uidByte, card.uid.uidByte, sizeof(card.uid.uidByte));
+	
+	if (result != STATUS_OK)
+		return false;
+
+	// RATS for SAK with sixth bit on.
+	if ((card.uid.sak & 0x20) == 0x20) {
+		byte atsBuffer[16];
+		byte atsBufferSize = 16;
+
+		result = PICC_RATS(atsBuffer, &atsBufferSize);
+		if (result == STATUS_OK) {
+			memset(card.ats, 0, sizeof(card.ats));
+			memcpy(card.ats, atsBuffer, atsBufferSize);
+
+			// I'm only taking care of MIFARE DESFire so the PPS
+			// Command is oriented to this type of card
+			PICC_Type piccType = PICC_GetType(&card);
+			if (piccType == PICC_TYPE_MIFARE_DESFIRE) {
+				result = PICC_PPS(0x00, 0x11, 0x00);
```

Comment:
That's a good question! I don't remember. 

I guess i was going to do the enhancements I mention on another comment just below... I think I really jumped a little just to get it working up to the DESFire code and the result was just left behind. I think that's the reason because:

    // I'm only taking care of MIFARE DESFire so the PPS
    // Command is oriented to this type of card
    PICC_Type piccType = PICC_GetType(&card);
    if (piccType == PICC_TYPE_MIFARE_DESFIRE) {

Really if the PPS is present it shoud be sent for ISO/IEC-14443 cards and not only DESFire. I guess I turned a bit lazy and instead of adding those check I went into a known condition where it is a DESFire. The PPS has also been hardcoded and therefore it is telling the card to use the same speed as the MFRC522 has setup by default. Therefore if it accepts the command there is no need to tell the MFRC522 to change it's speed and if it fails probablly PPS was not supported (That check should come from the ATS answer). i guess i just got lazy and left behind some clues of my lazyness. I'll look into it :)